File: 373. Find K Pairs with Smallest Sums/Min-heap-indices.py
from heapq import heappush, heappop

class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        m, n = len(nums1), len(nums2)
        k = min(k, m * n)
        result = []
        if not m or not n:
            return result
        
        pq = []
        for i in range(min(m, k)):
            heappush(pq, (nums1[i] + nums2[0], i, 0))
        
        while k > 0 and pq:
            _, i, j = heappop(pq)
            result.append([nums1[i], nums2[j]])
            k -= 1
            if j + 1 < n:
                heappush(pq, (nums1[i] + nums2[j + 1], i, j + 1))
        
        return result

File: 373. Find K Pairs with Smallest Sums/Min-heap-indices.java
class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums1.length == 0 || nums2.length == 0 || k == 0) return res;
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> (nums1[a[0]] + nums2[a[1]]) - (nums1[b[0]] + nums2[b[1]]));
        for (int j = 0; j < nums2.length && j < k; j++) {
            pq.offer(new int[]{0, j});
        }
        
        while (k-- > 0 && !pq.isEmpty()) {
            int[] curr = pq.poll();
            res.add(Arrays.asList(nums1[curr[0]], nums2[curr[1]]));
            if (curr[0] + 1 < nums1.length) {
                pq.offer(new int[]{curr[0] + 1, curr[1]});
            }
        }
        return res;
    }
}